### PR TITLE
update modular_modernbert -- add inputs_embeds param to ModernBertModel

### PR DESCRIPTION
**Pull request by jxmorris12**
# What does this PR do?
Hi! Congrats on the release of ModernBERT; it looks amazing. I'm interested in using ModernBERT eventually to train a new [Contextual Document Embeddings](https://arxiv.org/abs/2410.02525) model. 

One desired feature is to pass the contextual and word embeddings together in the second stage, which requires setting the `inputs_embeds` kwarg so that we can pass hidden states directly. This is a feature of typical BERT and other transformer implementations but isn't yet allowed by ModernBERT, so I added it. It's only a few additional lines of code.

cc: @warner-benjamin @tomaarsen @orionw @staghado @bclavie @NohTow @ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Yep would be nice to add. let's make it a bit simpler (si how llama has less if elses) and add a small test ! 🤗

**Comment by tomaarsen** [comment]
Hello @jxmorris12, @ArthurZucker,

I pushed some changes into this PR to get it closer to completion. Let me know if you're not okay with this, and you can easily revert or delete the commits.

The changes:
1. The `inputs_embeds` shouldn't fully replace `self.embeddings(input_ids)`, because this call also does layer normalization and dropout. So, now both `input_ids` and `inputs_embeds` is passed to the `ModernBertEmbeddings`, much like how `BertEmbeddings` is implemented.
2. I also added `inputs_embeds` to the docstring, and propagated the changes to the other model classes.
3. I also introduced an error if `input_ids` and `input_embeds` are both or neither provided.
4. I fixed an issue with device being based solely on `input_ids` with `attention_mask`.
5. Fix a test, and reintroduce another test.

> let's make it a bit simpler (si how llama has less if elses)

This is sadly not as simple as it seems due to `_unpad_modernbert_input`.

- Tom Aarsen

**Review by tomaarsen — APPROVED** [review]

**Comment by NohTow** [comment]
Hello,

Thanks for opening this PR, I really love your work on CDE and I look forward to seeing how it will perform with ModernBERT!
FYI, this feature (being able to pass input_embeds) has also been asked for other use cases [here](https://github.com/AnswerDotAI/ModernBERT/issues/165), so besides retro-compatibility and CDE, it seems like this feature is used by the community and is thus a very cool addition.

I'll make a more thorough review when I come back from vacation this week, but I already checked how to implement it for the linked issue and it seems to be in line with the latest change from Tom.

**Comment by jxmorris12** [comment]
> Hello @jxmorris12, @ArthurZucker,
> 
> I pushed some changes into this PR to get it closer to completion. Let me know if you're not okay with this, and you can easily revert or delete the commits.
...

thanks Tom! Looks great.

**Comment by jxmorris12** [comment]
Hi @NohTow @ArthurZucker just following up on this -- thanks :)

**Comment by znsoftm** [comment]
when will it be merged into the main branch?

File: src/transformers/models/modernbert/modular_modernbert.py [hunk 1]
```diff
@@ -1017,35 +1027,49 @@ def forward(
...
+
+        if input_ids is not None:
+            self.warn_if_padding_and_no_attention_mask(input_ids, attention_mask)
 
         if batch_size is None and seq_len is None:
-            batch_size, seq_len = input_ids.shape[:2]
+            if inputs_embeds is not None:
+                batch_size, seq_len = inputs_embeds.shape[:2]
+            else:
+                batch_size, seq_len = input_ids.shape[:2]
```
Comment (ArthurZucker) [inline src/transformers/models/modernbert/modular_modernbert.py:1045]: if you first embed input ids you don't need two branches!

File: src/transformers/models/modernbert/modular_modernbert.py [hunk 2]
```diff
@@ -1017,35 +1027,49 @@ def forward(
...
-                        inputs=input_ids, attention_mask=attention_mask
+                if inputs_embeds is None:
+                    with torch.no_grad():
+                        input_ids, indices, cu_seqlens, max_seqlen, *_ = _unpad_modernbert_input(
+                            inputs=input_ids, attention_mask=attention_mask
+                        )
+                else:
+                    inputs_embeds, indices, cu_seqlens, max_seqlen, *_ = _unpad_modernbert_input(
+                        inputs=inputs_embeds, attention_mask=attention_mask
                     )
```
Comment (ArthurZucker) [inline src/transformers/models/modernbert/modular_modernbert.py:1063]: only the input is different here! if you first embed the input, then you don't need two branches!

File: src/transformers/models/modernbert/modular_modernbert.py [hunk 3]
```diff
@@ -1201,19 +1226,32 @@ def forward(
 
         if self.config._attn_implementation == "flash_attention_2":
             if indices is None and cu_seqlens is None and max_seqlen is None:
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/modernbert/modular_modernbert.py:1228]: TBH it might make more sense! This way only the base model unpads, and the other models can freely do the unpadding. Kinda up to you 🤗

**Review by ArthurZucker — COMMENTED** [review]
Thanks for updating, let's remove complexity and good to go!

**Comment by NohTow** [comment]
Hello,
Sorry for the delay.
As mentioned earlier, the code looks good to me.
I only have two small nitpick. As mentioned by @ArthurZucker, first embedding the input_ids should be cleaner and less error prone.
The second is that the XOR test is somewhat a bit less informative than checking if the two are not set together and then checking if at least one is defined. Maybe it's just me, but I find `"You must specify exactly one of input_ids or inputs_embeds"` a bit ambiguous, yet I don't have a better wording.  

Besides those very small details, everything looks good!

File: src/transformers/models/modernbert/modular_modernbert.py (hunk 2 above)
Comment (jxmorris12) [reply src/transformers/models/modernbert/modular_modernbert.py:1063]: Well the old code contains a torch.no_grad here. That's why there are two branches. Did you all mean to wrap the unpad function in a no_grad? If so, we need to keep it. We definitely can't unpad inputs_embeds in a no_grad block because it will break gradient flow

**Comment by NohTow** [comment]
@jxmorris12 I am very sorry my message was unclear. I did not mean that the code was incorrect nor that testing the two conditions sequentially was cleaner, I just meant that from my reading, the error message could be ambiguous (in the case where the user do not feed anything, the _exactly_ might not be explicit enough).
But again, that is really a nitpick and probably comes from me not properly reading the message, let's ignore that!

**Comment by jxmorris12** [comment]
@jon-tow Oh okay, that makes sense to me! I agree it could be a really confusing error for the user. But maybe since it's in a lot of files (such as LLAMA) we could open a separate issue to improve that error message everywhere?

**Comment by NohTow** [comment]
@jxmorris12 Yeah totally, I raised that because I compared the implementations from BERT and this one and the BERT one does the check sequentially and thus have more informative messages, but since Arthur raised that it is already done like that in other models and is standard now, let's just follow the standard!

**Review by ArthurZucker — COMMENTED** [review]
My comments are not adressed 😓 
- let's embed inputs, then we computes shapes
- let's never require grad on unpadding / padding as it's weight agnostic

**Comment by NohTow** [comment]
For the latter, I think it is related to [this](https://github.com/huggingface/transformers/pull/35404).
It was discussed during the original PR, but it breaks the gradients.
I'll let @warner-benjamin give more information.

File: src/transformers/models/modernbert/modular_modernbert.py (hunk 2 above)
Comment (tomaarsen) [reply src/transformers/models/modernbert/modular_modernbert.py:1060]: I believe #35386 is a more detailed bug report on the gradient breaks

**Comment by ArthurZucker** [comment]
Got it! Okay, for output, we do need the gradients, my bad! 
TLDR: 
- when training: with torch no grad for input, but you need output gradients
- when training: with input embeddings: torch grad because you need propagation 
- when inference, never grad, but usually people do this outside the modeling. 

So the only "gain" is using no grad on input ids all the time

File: src/transformers/models/modernbert/modular_modernbert.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/modernbert/modular_modernbert.py:1060]: My bad! Indeed for input embedding you need grads if you are training an encoder!

**Review by ArthurZucker — APPROVED** [review]
Thanks for answering my questions! Let's go 🤗
